File: app/src/main/java/piuk/blockchain/android/ui/settings/customprefs/CustomSwitchPreferenceCompat.java
package piuk.blockchain.android.ui.settings.customprefs;

import android.content.Context;
import android.graphics.Typeface;
import android.support.v7.preference.SwitchPreferenceCompat;
import android.util.AttributeSet;

import uk.co.chrisjenx.calligraphy.CalligraphyUtils;
import uk.co.chrisjenx.calligraphy.TypefaceUtils;


public class CustomSwitchPreferenceCompat extends SwitchPreferenceCompat {

    private Typeface typeface;

    public CustomSwitchPreferenceCompat(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    public CustomSwitchPreferenceCompat(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public CustomSwitchPreferenceCompat(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public CustomSwitchPreferenceCompat(Context context) {
        super(context);
        init();
    }

    private void init() {
        typeface = TypefaceUtils.load(getContext().getAssets(), "fonts/Montserrat-Regular.ttf");
        // Forces setting fonts when Summary or Title are set via XMl
        setTitle(getTitle());
        setSummary(getSummary());
    }

    @Override
    public void setTitle(int titleResId) {
        setTitle(getContext().getString(titleResId));
    }

    @Override
    public void setTitle(CharSequence title) {
        CharSequence charSequence = CalligraphyUtils.applyTypefaceSpan(
                title,
                typeface);
        super.setTitle(charSequence);
    }

    @Override
    public void setSummary(int summaryResId) {
        setSummary(getContext().getString(summaryResId));
    }

    @Override
    public void setSummary(CharSequence summary) {
        CharSequence charSequence = CalligraphyUtils.applyTypefaceSpan(
                summary,
                typeface);
        super.setSummary(charSequence);
    }

}

File: app/src/main/java/piuk/blockchain/android/ui/settings/customprefs/CustomSwitchPreferenceCompat.kt
package piuk.blockchain.android.ui.settings.customprefs

import android.content.Context
import android.graphics.Typeface
import android.support.v7.preference.R
import android.support.v7.preference.SwitchPreferenceCompat
import android.util.AttributeSet
import piuk.blockchain.android.util.extensions.applyFont
import piuk.blockchain.android.util.helperfunctions.CustomFont
import piuk.blockchain.android.util.helperfunctions.loadFont

@Suppress("unused")
class CustomSwitchPreferenceCompat @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = R.attr.switchPreferenceCompatStyle,
        defStyleRes: Int = 0
) : SwitchPreferenceCompat(context, attrs, defStyleAttr, defStyleRes) {

    init {
        init()
    }

    private var typeface: Typeface? = null

    private fun init() {
        loadFont(context, CustomFont.MONTSERRAT_REGULAR) {
            typeface = it
            // Forces setting fonts when Summary or Title are set via XMl
            this.title = title
            this.summary = summary
        }
    }

    override fun setTitle(titleResId: Int) {
        title = context.getString(titleResId)
    }

    override fun setTitle(title: CharSequence?) {
        title?.let { super.setTitle(title.applyFont(typeface)) } ?: super.setTitle(title)
    }

    override fun setSummary(summaryResId: Int) {
        summary = context.getString(summaryResId)
    }

    override fun setSummary(summary: CharSequence?) {
        summary?.let { super.setSummary(summary.applyFont(typeface)) } ?: super.setSummary(summary)
    }

}
